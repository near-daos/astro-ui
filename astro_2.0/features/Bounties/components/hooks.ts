import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
