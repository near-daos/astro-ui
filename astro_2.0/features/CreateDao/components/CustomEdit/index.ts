import CustomEdit from './CustomEdit';

export default CustomEdit;
