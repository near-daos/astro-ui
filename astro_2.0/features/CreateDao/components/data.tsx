import {
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
  DaoSubjectInfo,
} from 'astro_2.0/features/CreateDao/components/types';

const DAO_PROPOSALS_OPEN: DaoSettingOption<DAOProposalsType> = {
  value: 'open',
  icon: 'illustrationOpenOrganization',
  subject: 'proposals',
  title: 'daoRulesAll',
  description: 'daoRulesAllDescription',
};

const DAO_PROPOSALS_CLOSED: DaoSettingOption<DAOProposalsType> = {
  value: 'closed',
  icon: 'illustrationMembersOnly',
  subject: 'proposals',
  title: 'daoRulesMembers',
  description: 'daoRulesMembersDescription',
};

const DAO_VOTING_POWER_DEMOCRATIC: DaoSettingOption<DAOVotingPowerType> = {
  value: 'democratic',
  icon: 'illustrationVotePerMember',
  subject: 'voting',
  title: 'daoRulesDemocratic',
  description: 'daoRulesDemocraticDescription',
};

const DAO_VOTING_POWER_TOKEN: DaoSettingOption<DAOVotingPowerType> = {
  value: 'weighted',
  icon: 'illustrationTokenWeighted',
  subject: 'voting',
  title: 'daoRulesTokenWeighted',
  description: 'daoRulesTokenWeightedDescription',
};

const DAO_STRUCTURE_FLAT: DaoSettingOption<DAOStructureType> = {
  value: 'flat',
  icon: 'illustrationFlatOrganization',
  subject: 'structure',
  title: 'daoRulesFlat',
  description: 'daoRulesFlatDescription',
};

const DAO_STRUCTURE_GROUPS: DaoSettingOption<DAOStructureType> = {
  value: 'groups',
  icon: 'illustrationGroupsAndCommitttees',
  subject: 'structure',
  title: 'daoRulesGroups',
  description: 'daoRulesGroupsDescription',
};

export const DAO_PROPOSALS_OPTIONS: Record<
  DAOProposalsType,
  DaoSettingOption<DAOProposalsType>
> = {
  open: DAO_PROPOSALS_OPEN,
  closed: DAO_PROPOSALS_CLOSED,
};

export const DAO_VOTING_POWER_OPTIONS: Record<
  DAOVotingPowerType,
  DaoSettingOption<DAOVotingPowerType>
> = {
  democratic: DAO_VOTING_POWER_DEMOCRATIC,
  weighted: DAO_VOTING_POWER_TOKEN,
};

export const DAO_STRUCTURE_OPTIONS: Record<
  DAOStructureType,
  DaoSettingOption<DAOStructureType>
> = {
  flat: DAO_STRUCTURE_FLAT,
  groups: DAO_STRUCTURE_GROUPS,
};

export const DAO_SUBJECT_OPTIONS = {
  voting: Object.values(DAO_VOTING_POWER_OPTIONS),
  structure: Object.values(DAO_STRUCTURE_OPTIONS),
  proposals: Object.values(DAO_PROPOSALS_OPTIONS),
};

export const DAO_RULES_INFO: DaoSubjectInfo[] = [
  {
    subject: 'proposals',
    title: 'daoRulesProposals',
    subTitle: 'daoRulesProposalsDescription',
  },
  {
    subject: 'voting',
    title: 'daoRulesVoting',
    subTitle: 'daoRulesVotingDescription',
  },
  {
    subject: 'structure',
    title: 'daoRulesStructure',
    subTitle: 'daoRulesStructureDescription',
  },
];
