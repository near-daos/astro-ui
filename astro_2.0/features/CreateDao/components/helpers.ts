import { GlobalState } from 'little-state-machine';
import {
  AssetsStep,
  GroupsStep,
  InfoStep,
  LegalStep,
  LinksStep,
  MembersStep,
  ProposalsStep,
  SubmitStep,
  VotingStep,
} from 'astro_2.0/features/CreateDao/types';
import { AnySchema } from 'yup';
import { ResolverError, ResolverResult } from 'react-hook-form';
import { DEFAULT_CREATE_DAO_GAS } from 'services/sputnik/constants';

export function getInitialValues(
  accountId: string,
  defaultFlag: string
): GlobalState {
  return {
    info: {
      displayName: '',
      address: '',
      purpose: '',
      isValid: false,
    },
    kyc: {
      legalStatus: '',
      legalLink: '',
      isValid: false,
    },
    links: {
      websites: [],
      isValid: false,
    },
    groups: {
      items: [{ name: 'council', slug: 'council' }],
      isValid: false,
    },
    members: {
      accounts: [],
      isValid: false,
    },
    assets: {
      flagCover: '',
      flagLogo: '',
      defaultFlag,
      isValid: false,
    },
    voting: {
      data: null,
      isValid: false,
    },
    proposals: {
      data: null,
      isValid: false,
    },
    submit: {
      isValid: true,
      gas: DEFAULT_CREATE_DAO_GAS,
    },
  };
}

export function updateAction(
  state: GlobalState,
  payload: Record<
    string,
    | InfoStep
    | LegalStep
    | LinksStep
    | MembersStep
    | GroupsStep
    | AssetsStep
    | VotingStep
    | ProposalsStep
    | SubmitStep
  >
): GlobalState {
  return {
    ...state,
    ...payload,
  };
}

export async function handleValidate<T>(
  schema: AnySchema,
  data: T,
  onUpdateCb: (valid: boolean, data?: T, errors?: ResolverError<T>) => void
): Promise<ResolverResult<T>> {
  try {
    const values = await schema.validate(data, {
      abortEarly: false,
    });

    onUpdateCb(true);

    return {
      values,
      errors: {},
    };
  } catch (e) {
    onUpdateCb(false);

    return {
      values: {},
      errors: e.inner.reduce(
        (
          allErrors: Record<string, string>,
          currentError: { path: string; type?: string; message: string }
        ) => {
          return {
            ...allErrors,
            [currentError.path]: {
              type: currentError.type ?? 'validation',
              message: currentError.message,
            },
          };
        },
        {}
      ),
    };
  }
}
