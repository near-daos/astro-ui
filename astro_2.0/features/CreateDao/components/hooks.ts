import {
  useFormContext,
  UseFormGetValues,
  UseFormSetValue,
} from 'react-hook-form';
import {
  DAOFormValues,
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
} from 'astro_2.0/features/CreateDao/components/types';
import { useCallback, useEffect, useState } from 'react';
import {
  DAO_PROPOSALS_OPTIONS,
  DAO_STRUCTURE_OPTIONS,
  DAO_VOTING_POWER_OPTIONS,
} from 'astro_2.0/features/CreateDao/components/data';
import { httpService } from 'services/HttpService';
import { CreateDaoCustomInput, CreateDaoInput } from 'types/dao';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import { SputnikWalletError } from 'errors/SputnikWalletError';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import {
  getInitialValues,
  updateAction,
} from 'astro_2.0/features/CreateDao/components/helpers';
import { useStateMachine } from 'little-state-machine';
import { useWalletContext } from 'context/WalletContext';
import { GA_EVENTS, sendGAEvent } from 'utils/ga';

type DaoFormStateReturn = {
  options: DaoSettingOption<
    DAOStructureType | DAOVotingPowerType | DAOProposalsType
  >[];
  getValues: UseFormGetValues<DAOFormValues>;
  setValue: UseFormSetValue<DAOFormValues>;
};

export const useDaoFormState = (): DaoFormStateReturn => {
  const { getValues, watch, setValue } = useFormContext<DAOFormValues>();

  const [structure, voting, proposals] = watch([
    'structure',
    'voting',
    'proposals',
  ]);

  const [options, setOptions] = useState<
    DaoSettingOption<DAOStructureType | DAOVotingPowerType | DAOProposalsType>[]
  >([]);

  useEffect(() => {
    setOptions([
      DAO_STRUCTURE_OPTIONS[structure],
      DAO_VOTING_POWER_OPTIONS[voting],
      DAO_PROPOSALS_OPTIONS[proposals],
    ]);
  }, [structure, voting, proposals, setValue]);

  return { options, getValues, setValue };
};

export function useImageUpload(): {
  uploadImage: (file: File) => Promise<string>;
} {
  const uploadImage = useCallback(async (img: File) => {
    try {
      if (img) {
        const { data } = await httpService.post('/api/upload-to-s3', img, {
          baseURL: '',
        });

        return data;
      }
    } catch (e) {
      return '';
    }

    return '';
  }, []);

  return { uploadImage };
}

export function useCreateDao(): {
  uploadAssets: (defaultFlag: string) => Promise<string>;
  createDao: (
    daoName: string,
    data: CreateDaoInput | CreateDaoCustomInput
  ) => Promise<void>;
} {
  const router = useRouter();
  const { t } = useTranslation();
  const { accountId, nearService } = useWalletContext();
  const { actions, state } = useStateMachine({ updateAction });

  const uploadImg = useCallback(async (img: File) => {
    if (img) {
      const { data } = await httpService.post('/api/upload-to-s3', img, {
        baseURL: '',
      });

      return data;
    }

    return '';
  }, []);

  async function loadImage(defaultFlag: string) {
    const response = await fetch(defaultFlag);
    const blob = await response.blob();

    return new File([blob], 'image.png', { type: 'image/png' });
  }

  const uploadAssets = useCallback(
    async (defaultFlag: string) => {
      try {
        const defaultFlagFile = await loadImage(defaultFlag);

        return uploadImg(defaultFlagFile);
      } catch (e) {
        return '';
      }
    },
    [uploadImg]
  );

  const createDao = useCallback(
    async (daoName: string, data: CreateDaoInput | CreateDaoCustomInput) => {
      try {
        await nearService?.createDao(data);

        const { nearConfig } = configService.get();
        const daoId = `${daoName}.${nearConfig?.contractName ?? ''}`;

        sendGAEvent({
          name: GA_EVENTS.CREATE_DAO,
          daoId,
          accountId,
        });

        showNotification({
          type: NOTIFICATION_TYPES.INFO,
          description: t('notifications.transactionDelay'),
          lifetime: 20000,
        });

        sessionStorage.removeItem('__LSM__');
        actions.updateAction(
          getInitialValues(accountId, state.assets.defaultFlag)
        );

        await router.push(`/dao/${daoId}`);
      } catch (error) {
        console.warn(error);

        if (error instanceof SputnikWalletError) {
          showNotification({
            type: NOTIFICATION_TYPES.ERROR,
            description: error.message,
            lifetime: 20000,
          });
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [router, uploadImg, nearService]
  );

  return {
    createDao,
    uploadAssets,
  };
}
