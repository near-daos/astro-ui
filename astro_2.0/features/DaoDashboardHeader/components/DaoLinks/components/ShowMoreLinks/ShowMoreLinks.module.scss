@import 'styles/mixins';
@import 'styles/typography';

.showMore {
  @extend %title5;
  @include centralize;
  background-color: var(--color-neutral-20);
  border-radius: 4px;
  height: 40px;
  margin-bottom: 0;
  width: 40px;

  @include tablet-landscape {
    border: 0;
    padding: 0;
  }

  &:hover {
    background-color: var(--color-neutral-30);
  }
}

.toggle {
  width: 24px;
}
