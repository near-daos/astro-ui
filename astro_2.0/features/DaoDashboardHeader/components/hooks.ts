import { useEffect, useState } from 'react';
import { SputnikHttpService } from 'services/sputnik';
import { UserPermissions } from 'types/context';
import { useAuthContext } from 'context/AuthContext';

type JoinDaoState = {
  showButton: boolean;
  showWarning: boolean;
};

export function useJoinDao(
  daoId: string,
  userPermissions: UserPermissions,
  daoMembers: string[]
): JoinDaoState {
  const { accountId } = useAuthContext();
  const [state, setState] = useState<JoinDaoState>({
    showButton: false,
    showWarning: false,
  });

  useEffect(() => {
    (async () => {
      if (
        !userPermissions.isCanCreateProposals ||
        daoMembers.includes(accountId)
      ) {
        return;
      }

      const pendingJoinApproval = await SputnikHttpService.getJoiningDaoProposal(
        { daoId, accountId }
      );

      if (pendingJoinApproval) {
        setState({
          showButton: false,
          showWarning: true,
        });

        return;
      }

      setState({
        showButton: true,
        showWarning: false,
      });
    })();
  }, [accountId, daoId, daoMembers, userPermissions.isCanCreateProposals]);

  return state;
}
