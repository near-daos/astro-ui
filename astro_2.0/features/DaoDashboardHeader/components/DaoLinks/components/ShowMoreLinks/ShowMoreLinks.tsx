import isEmpty from 'lodash/isEmpty';
import React, { useCallback } from 'react';
import { useModal } from 'components/modal';
import { MoreLinksModal } from 'astro_2.0/features/DaoDashboardHeader/components/DaoLinks/components/MoreLinksModal/MoreLinksModal';
import { Button } from 'components/button/Button';
import { Icon } from 'components/Icon';

import styles from './ShowMoreLinks.module.scss';

interface ShowMoreLinksProps {
  links: string[];
}

export const ShowMoreLinks: React.FC<ShowMoreLinksProps> = ({ links }) => {
  const [showModal] = useModal(MoreLinksModal, { links });
  const handleClick = useCallback(async () => {
    await showModal();
  }, [showModal]);

  if (isEmpty(links)) {
    return null;
  }

  return (
    <Button
      size="block"
      variant="transparent"
      onClick={handleClick}
      className={styles.showMore}
    >
      <Icon name="buttonLink" className={styles.toggle} />
    </Button>
  );
};
