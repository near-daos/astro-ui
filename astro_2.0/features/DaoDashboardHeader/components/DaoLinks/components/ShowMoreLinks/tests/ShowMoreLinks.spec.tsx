import { render } from 'jest/testUtils';

import { useModal } from 'components/modal';

import { ShowMoreLinks } from 'astro_2.0/features/DaoDashboardHeader/components/DaoLinks/components/ShowMoreLinks';
import { fireEvent } from '@testing-library/dom';

jest.mock('components/modal', () => {
  return {
    useModal: jest.fn(() => []),
  };
});

describe('ShowMoreLinks', () => {
  it('Should render nothing', () => {
    const { container } = render(<ShowMoreLinks links={[]} />);

    expect(container).toMatchSnapshot();
  });

  it('Should call showModal', () => {
    const showModal = jest.fn();

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useModal.mockImplementation(() => [showModal]);

    const { getByRole } = render(<ShowMoreLinks links={['L1']} />);

    fireEvent.click(getByRole('button'));

    expect(showModal).toBeCalled();
  });
});
