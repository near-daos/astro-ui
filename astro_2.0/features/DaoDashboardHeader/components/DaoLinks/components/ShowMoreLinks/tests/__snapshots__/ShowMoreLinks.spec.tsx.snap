// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMoreLinks Should render nothing 1`] = `<div />`;
